src/main/scala/viper/gobra/ast/internal/transform/OverflowChecksTransform.scala

```diff
@@ -0,0 +1,142 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.ast.internal.transform
+
+import viper.gobra.ast.internal._
+import viper.gobra.reporting.Source
+import viper.gobra.reporting.Source.Parser.Single
+import viper.gobra.util.TypeBounds.BoundedIntegerKind
+import viper.gobra.util.Violation.violation
+
+/**
+  * Adds overflow checks to programs written in Gobra's internal language
+  */
+object OverflowChecksTransform extends InternalTransform {
+  override def name(): String = "add_integer_overflow_checks"
+
+  override def transform(p: Program): Program = transformMembers(memberTrans)(p)
+
+  private def memberTrans(member: Member): Member = member match {
+    // Adds overflow checks per statement that contains an expression and assume statements in the beginning
+    // specifying the bounds of each argument
+    case f@Function(name, args, results, pres, posts, body) =>
+      Function(name, args, results, pres, posts, body map (computeNewBody(args, _)))(f.info)
+
+    // same as functions
+    case m@Method(receiver, name, args, results, pres, posts, body) =>
+      Method(receiver, name, args, results, pres, posts, body map (computeNewBody(args, _)))(m.info)
+
+    // Adds pre-conditions stating the bounds of each argument and a post-condition to check if the body expression
+    // overflows
+    case f@PureFunction(name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPre(args)
+        val newPost = posts ++ Vector(getPureBlockPost(expr, results))
+        PureFunction(name, args, results, newPre, newPost, body)(f.info)
+      case None => f
+    }
+
+    // Same as pure functions
+    case m@PureMethod(receiver, name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPre(args)
+        val newPost = posts ++ Vector(getPureBlockPost(expr, results))
+        PureMethod(receiver, name, args, results, newPre, newPost, body)(m.info)
+      case None => m
+    }
+
+    // TODO: should we add overflow checks to predicates?
```

Comment:
Let's discuss this today in the Gobra meeting

Reply:
Added a small comment about this on 46a41771680dbcbaf8764875afa5e8a9d084bd01